### ISSUE TITLE
Why The Monad Pattern Looks Promising in Go+Torch Design

### ISSUE THREAD

**Issue by shendiaomo** [wangkuiyi/gotorch#3]
Monad is a programming pattern that records the output of each function call in a data structure, so we can free them at once afterward.  It applies to many programming languages.  Let us see why it is important to Go+Torch.

Go uses the pattern extensively, see https://www.innoq.com/en/blog/golang-errors-monads/ for an example.

## Case Study 1: Free Tensors

We now allocate `Tensor` objects using `new` to keep the reference count in the `shared_ptr` field of the C++ `Tensor` class: https://github.com/wangkuiyi/gotorch/blob/4ade9aa9ce84ae2532df260dd910c2b7bcf1e47a/cgotorch/cgotorch.cc#L16
The `Tensor` objects `new`ed would cause memory leak if we don't recycle them.

Assume that Go has a similar frontend API as C++, then according to the [C++ MNIST example](https://github.com/pytorch/examples/blob/8df8e747857261ea481e0b2492413d52bf7cc3a8/cpp/mnist/mnist.cpp), let's think about the following problems.
#### Problem 1: Destruct `Tensor`s Created In The Train Loop To Avoid Memory Leak
1. `Tensor`s Created In the [C++ train loop](https://github.com/pytorch/examples/blob/8df8e747857261ea481e0b2492413d52bf7cc3a8/cpp/mnist/mnist.cpp#L65-L82):
    The train loop in mnist.cpp is like:
    ```c++
    for (auto& batch : data_loader) {
        auto data = batch.data.to(device), targets = batch.target.to(device);  // `data` and `targets` are `Tensor`s
        optimizer.zero_grad();
        auto output = model.forward(data);  // `output` is a `Tensor`
        auto loss = torch::nll_loss(output, targets);  // `loss` is a `Tensor`
        AT_ASSERT(!std::isnan(loss.template item<float>()));
        loss.backward();
        optimizer.step();
        //...
    }
    ```
    We can see these `Tensor`s has to be created:
    1. `data` and `targets` as the features and labels of the dataset
    2. `output` as the predictions of the `data`
    3. `loss` 

    #### We can use `defer` to destruct `Tensor`s in the train loop
    Because `data`, `targets`, `output`, and `loss` are all stack variables, they are created and destroyed in each iteration of the C++ train loop. **This implied the *libtorch* framework would take ownership of the `Tensor` s if necessary.** As a result, a naive API of `gotorch` can use `defer` to recycle the reference-counted `Tensor`s. That is, the following imaginary code would work okay.
    ```go
    // We need this nested function to make `defer` works as expected.
    func step(batch *Batch) {
        // `data`, `targets`, `output`, `loss` are `Tensor`s.
        data := batch.Data.To(device)
        defer data.Close()
        target := batch.Target.To(device)  
        defer target.Close()
        optimizer.zero_grad()
        output := model.Forward(data)
        defer output.Close()
        loss = torch.NllLoss(output, targets)
        defer loss.Close()
        loss.Backward()
        optimizer.Step()
        // ...
    }
    for batch := range data_loader {
        step(batch)
    }
    ```
    The `defer`s are a bit tedious, maybe we can improve the syntax of Go+ to save typing.
2. `Tensor`s Created In the [C++ `forward` method](https://github.com/pytorch/examples/blob/master/cpp/mnist/mnist.cpp#L37-L46)
    The `forward` method is called by the train loop above, in the C++ mnist example, the `forward` method looks like:
    ```c++
    torch::Tensor forward(torch::Tensor x) {
        x = torch::relu(torch::max_pool2d(conv1->forward(x), 2));
        x = torch::relu(
            torch::max_pool2d(conv2_drop->forward(conv2->forward(x)), 2));
        x = x.view({-1, 320});
        x = torch::relu(fc1->forward(x));
        x = torch::dropout(x, /*p=*/0.5, /*training=*/is_training());
        x = fc2->forward(x);
        return torch::log_softmax(x, /*dim=*/1);
      }
    ```
    #### We can use `defer` to destruct `Tensor`s in the `Forward` function (in a tricky way)
    Similar to the train loop above, `x` is a `Tensor` on the stack and is destroyed at the end of the function scope. The difference is that `x` is reassigned multiple times. So we cannot simply use `defer x.Close()` here. A workaround is requiring users to use a different idiom, for a naive example:
    ```golang
    func (net *Net) Forward(x torch.Tensor) torch.Tensor {  // The argument x is recycled in the train loop
        var tensors []Tensor
        defer func () {
            for t := range tensors {
                t.Close()
            }
        }()
        x = torch.Relu(torch.MaxPool2d(net.conv1.Forward(x), 2))
        append(tensors, x)
        x = torch.Relu(
            torch.MaxPool2d(net.conv2_drop.Forward(net.conv2.Forward(x)), 2))
        append(tensors, x)
        x = x.View([]int{-1, 320})
        append(tensors, x)
        x = torch.Relu(net.fc1.Forward(x))
        append(tensors, x)
        x = torch.Dropout(x, /*p=*/0.5, /*training=*/is_training())
        append(tensors, x)
        x = net.fc2.Forward(x)
        append(tensors, x)
        return torch.LogSoftmax(x, /*dim=*/1)  // The return value is recycled in the train loop
      }
    ```
    Obviously, this is not very elegant. 
    #### Should we bookkeeping the `Tensors` in C++?
    A better way is keeping the `tensors` array in C++ rather than in Go, for example, we can use `std::vector` to record each C++ `Tensor` created by Go API, and provide a `torch.CleanTensors` for users to call at the end of the train loop. However, this solution is harder to design properly, for example, we have to take `goroutine`s into consideration so as to avoid corrupting the `std::vector`.

## Case Study 2: Record Errors

Few functions in *libtorch* have the `noexcept` tag. This implies that most of the functions in C++ may throw an exception. We have to expose an `error` return type for these functions' wrappers in Go. Recall the `step` function above:
```golang
func step(batch *Batch) {
    // `data`, `targets`, `output`, `loss` are `Tensor`s.
    data := batch.Data.To(device)
    defer data.Close()
    // ...
}
```
It may become the following in production code:
```golang
func step(batch *Batch) error {
    // `data`, `targets`, `output`, `loss` are `Tensor`s.
    data, err := batch.Data.To(device)
    if err != nil {
        return ...
    }
    defer data.Close()
    // ...
}
```
That is, the user should check whether there's an error on each line. This may be tedious too. Go+ has a neat syntax to [unwrap errors](https://github.com/goplus/gop/blob/master/tutorial/15-ErrWrap/err_wrap.gop), but I cannot think of an elegant way to solve the problem for the time being. See previous discussions also: https://github.com/goplus/gop/issues/307#issuecomment-663396846, https://github.com/goplus/gop/issues/307#issuecomment-663942929

**Comment by shendiaomo** [comment]
### Proposal: Use The Monad Pattern To Ease Error Handling And Memory Management
Monad is a concept coined by functional programming to reduce boilerplate code. Go uses the pattern extensively, see https://www.innoq.com/en/blog/golang-errors-monads/ for an example.

#### Example 0: The Monadic API
In a monadic pattern, we have to define a new type that encapsulates the *libtorch* functions. The type `Torch` has two fields `err` and `tensors` to reduce boilerplate code about memory management and error handling.
```golang
type Torch struct {
    err error
    tensors []Tensor
}

func (torch *Torch) Error() error {
    return torch.err
}

func (torch *Torch) Reclaim()  {
    for t := torch.tensors {
        Delete(t)  // `Delete` is a imaginary cgo wrapper function that calls C++ delete
    }
    torch.tensors = []Tensor{}
}

func (torch *Torch) Relu(tensor Tensor) Tensor {
    if torch.err != nil {
        return Tensor{}
    }
    t, err := relu(tensor) // `relu` is the imaginary cgo wrapper function
    if err != nil {
        torch.err = err // Record line number etc.
        return Tensor{}
    }
    append(torch.tensors, t) // Register `Tensor`s for later reclaiming
    return t
}

// Define `MaxPool2d`, `Dropout`, and other functions with a similar pattern as `Relu`
```
With the monadic API, a user-define `Module` struct should embed a `Torch`  object. As a result, users can write much neater code for their own `Module`s:

#### Example 1: The Forward Function
```golang
struct Net {
    torch.Torch
    // Define conv1, conv2, fc1, fc2 etc.
}

func (net *Net) Forward(x torch.Tensor) torch.Tensor {
    x = net.Relu(torch.MaxPool2d(net.conv1.Forward(x), 2))
    x = torch.Relu(
        torch.MaxPool2d(net.conv2_drop.Forward(net.conv2.Forward(x)), 2))
    x = x.View([]int{-1, 320})
    x = torch.Relu(net.fc1.Forward(x))
    x = torch.Dropout(x, /*p=*/0.5, /*training=*/is_training())
    x = net.fc2.Forward(x)
    return torch.LogSoftmax(x, /*dim=*/1)
```
As we can see, the code is much shorter than the same example in the original post.
NOTE: the users also have to pass the `tensors` slice of `Torch` to submodules such as `net.conv1`, `net.conv2` etc. in some way to register `Tensor`s for later reclaiming.
#### Example 2: The Train Loop
```golang
// We need this nested function to make `defer` works as expected.
func step(batch *Batch) {
    // `model` is an instance of  the `Net` struct above
    defer model.Reclaim() // `delete` all tensors at the end of scope as the C++ version
    data := model.ToDevice(batch.Data, device)
    target := model.ToDevice(batch.Target, device)
    optimizer.zero_grad()
    output := model.Forward(data)
    loss = model.NllLoss(output, targets)
    loss.Backward()
    optimizer.Step()
    // ...
}

for batch := range data_loader {
    step(batch)
    if model.err != nil {  // Use the monad API to handle errors
        return ...
    }
}
```
As we can see, the code is also much shorter than the same example in the original post and contains a complete error handling.

**Comment by shendiaomo** [comment]
Another way is to define a general monad，like:
struct Monad {
    err error
    tensors []Tensor
}

fund (m *Monad) Do(func f() Tensor) Tensor {
    if m.err != nil {
        
    }
}

**Comment by xushiwei** [comment]
Many be we can [introduce smart pointer in Go+](https://github.com/goplus/gop/issues/512)

**Comment by wangkuiyi** [comment]
We finally chose the `SetTensorFinalizer` solution, other than the Monad pattern.

**Comment by xushiwei** [comment]
> We finally chose the `SetTensorFinalizer` solution, other than the Monad pattern.

Where can I refer the detail about `SetTensorFinalizer` solution?

**Comment by wangkuiyi** [comment]
@xushiwei It is in https://github.com/wangkuiyi/gotorch/blob/develop/tensor_gc.go
